SDK Doc Update

appilon opened the pull request:
This PR updates all import paths, code links, and godoc.org links under the `extend` section to the new SDK location. Code links that were locked to a specific sha were switched to reference master for now, it's possible the line numbers aren't correct anymore so we should double check those.

I also wrote a blurb on the new SDK in a new sidebar item/page.

appilon committed:
update docs to reflect new import path
begin docs on SDK changeover

nfagerlund commented on content/source/docs/extend/best-practices/testing.html.md:
```diff
@@ -376,8 +376,8 @@ for safely managing infrastructure.
 [5]: /docs/extend/testing/acceptance-tests/teststep.html
-[6]: https://github.com/hashicorp/terraform/blob/9441e78fb9c35037da71fd0284b97e546dd6a53b/helper/resource/testing.go#L299-L301
-[7]:  https://github.com/hashicorp/terraform/blob/9441e78fb9c35037da71fd0284b97e546dd6a53b/helper/resource/testing.go#L303-L306
+[6]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L299-L301
```
```suggestion
[6]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L318-L320
```

nfagerlund commented on content/source/docs/extend/best-practices/testing.html.md:
```diff
@@ -376,8 +376,8 @@ for safely managing infrastructure.
-[6]: https://github.com/hashicorp/terraform/blob/9441e78fb9c35037da71fd0284b97e546dd6a53b/helper/resource/testing.go#L299-L301
-[7]:  https://github.com/hashicorp/terraform/blob/9441e78fb9c35037da71fd0284b97e546dd6a53b/helper/resource/testing.go#L303-L306
+[6]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L299-L301
+[7]:  https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L303-L306
```
```suggestion
[7]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L322-L325
```

nfagerlund commented on content/source/docs/extend/schemas/index.html.md:
```diff
@@ -49,7 +49,7 @@ In this example we’re creating a `Provider` and setting it’s `schema`. This
 user can specify in their configuration. The keys are strings, and the values
 are
-[`schema.Schema`](https://github.com/hashicorp/terraform/blob/5727d3335247e5940af2bef35c88657753f6d260/helper/schema/schema.go#L37)
+[`schema.Schema`](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L37)
```
```suggestion
[`schema.Schema`](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L61)
```

nfagerlund commented on content/source/docs/extend/schemas/schema-behaviors.html.md:
```diff
@@ -190,7 +190,7 @@ resource "example_instance" "ex" {
 ### DiffSuppressFunc
 **Data structure:**
-[SchemaDiffSuppressFunc](https://github.com/hashicorp/terraform/blob/ead558261d5e322f1f1e90c8e74834ba9215f24e/helper/schema/schema.go#L202)    
+[SchemaDiffSuppressFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L202)    
```
```suggestion
[SchemaDiffSuppressFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L263-L268)    
```

nfagerlund commented on content/source/docs/extend/schemas/schema-behaviors.html.md:
```diff
@@ -228,7 +228,7 @@ resource "example_instance" "ex" {
 ### DefaultFunc
 **Data structure:**
-[SchemaDefaultFunc](https://github.com/hashicorp/terraform/blob/ead558261d5e322f1f1e90c8e74834ba9215f24e/helper/schema/schema.go#L209)  
+[SchemaDefaultFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L209)  
```
```suggestion
[SchemaDefaultFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L270-L272)  
```

nfagerlund commented on content/source/docs/extend/schemas/schema-behaviors.html.md:
```diff
@@ -292,7 +292,7 @@ provider "example" {
 
 ### StateFunc
-**Data structure:** [SchemaStateFunc](https://github.com/hashicorp/terraform/blob/a20dbb43782ade816baaeffa8033da0027ee6b26/helper/schema/schema.go#L245)    
+**Data structure:** [SchemaStateFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L245)    
```
```suggestion
**Data structure:** [SchemaStateFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L306-L308)    
```

nfagerlund commented on content/source/docs/extend/schemas/schema-behaviors.html.md:
```diff
@@ -327,7 +327,7 @@ Value in statefile:
 
 ### ValidateFunc
-**Data structure:** [SchemaValidateFunc](https://github.com/hashicorp/terraform/blob/a20dbb43782ade816baaeffa8033da0027ee6b26/helper/schema/schema.go#L249)   
+**Data structure:** [SchemaValidateFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L249)   
```
```suggestion
**Data structure:** [SchemaValidateFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/schema.go#L310-L312)   
```

nfagerlund commented on content/source/docs/extend/schemas/schema-types.html.md:
```diff
@@ -290,7 +290,7 @@ resource "example_compute_instance" "ex" {
 
 ### TypeSet
-**Data structure:** [`*schema.Set`](https://github.com/hashicorp/terraform/blob/504d5ef233230984ccd378a6847fa8b9e32f6bee/helper/schema/set.go#L44)  
+**Data structure:** [`*schema.Set`](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/set.go#L44)  
```
```suggestion
**Data structure:** [`*schema.Set`](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/schema/set.go#L50)  
```

nfagerlund commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
````diff
@@ -54,7 +54,23 @@ packages accordingly. Once your provider codebase is a Go Module, you can
+```
+
+### Switching to the official SDK
+While the `helper` package has served us well, in order for provider development to evolve the SDK needed to break out into its own repository. Terraform v0.12.7 will be the final release with importable packages intended for provider development. Going forward developers will import [github.com/hashicorp/terraform-plugin-sdk](https://github.com/hashicorp/terraform-plugin-sdk). The first release  ...
````
```suggestion
While the `helper` package has served us well, in order for provider development to evolve the SDK needed to break out into its own repository. Terraform v0.12.7 will be the final release with importable packages intended for provider development. Going forward developers will import [github.com/hashicorp/terraform-plugin-sdk](https://github.com/hashicorp/terraform-plugin-sdk). The first release of this SDK aims to keep nearly 100% backwards compatibility, aside from a handful of APIs, so only the imports within your provider need to be replaced. The migration process can be automated with the [migrator tool](https://github.com/hashicorp/tf-sdk-migrator).
```

nfagerlund commented on content/source/docs/extend/testing/acceptance-tests/testcase.html.md:
```diff
@@ -143,7 +143,7 @@ func testAccPreCheck(t *testing.T) {
 
 **Type:**
-`map[string]`[terraform.ResourceProvider](https://github.com/hashicorp/terraform/blob/0cc9e050ecd4a46ba6448758c2edc0b29bef5695/terraform/resource_provider.go#L10-L171)  
+`map[string]`[terraform.ResourceProvider](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/terraform/resource_provider.go#L10-L171)  
```
```suggestion
`map[string]`[terraform.ResourceProvider](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/terraform/resource_provider.go#L12-L172)  
```

nfagerlund commented on content/source/docs/extend/testing/acceptance-tests/testcase.html.md:
```diff
@@ -186,7 +186,7 @@ func init() {
 ### CheckDestroy 
 
-**Type:** [TestCheckFunc](https://github.com/hashicorp/terraform/blob/0cc9e050ecd4a46ba6448758c2edc0b29bef5695/helper/resource/testing.go#L182-L186)  
+**Type:** [TestCheckFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L182-L186)  
```
```suggestion
**Type:** [TestCheckFunc](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L192-L196)  
```

nfagerlund commented on content/source/docs/extend/testing/acceptance-tests/testcase.html.md:
```diff
@@ -252,7 +252,7 @@ func testAccCheckExampleResourceDestroy(s *terraform.State) error {
 ### Steps 
 
-**Type:** [[]TestStep](https://github.com/hashicorp/terraform/blob/0cc9e050ecd4a46ba6448758c2edc0b29bef5695/helper/resource/testing.go#L249-L367)  
+**Type:** [[]TestStep](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L249-L367)  
```
```suggestion
**Type:** [`[]TestStep`](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L259-L391)  
```

nfagerlund commented on content/source/docs/extend/testing/acceptance-tests/testcase.html.md:
```diff
@@ -301,6 +301,6 @@ creation to creating, adding, and removing specific properties. In the next
 to create these scenarios by iterating on Terraform configurations. 
 
-[1]: https://github.com/hashicorp/terraform/blob/0cc9e050ecd4a46ba6448758c2edc0b29bef5695/helper/resource/testing.go#L195-L247
+[1]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L195-L247
```
```suggestion
[1]: https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/testing.go#L205-L257
```

nfagerlund reviewed (commented):
I helped out a bit with the line numbers. BTW, I strongly suggest linking to a specific sha like 7845463d4f2d789b2b066e4e15fde1d5376ebeb9 (which is what I just tested against) instead of master. I know falling behind head is annoying, but at least it preserves information about what we _thought_ we were linking to at the time we wrote it, which greatly simplifies any future updates.

appilon commented:
@nfagerlund Thanks for the help with the line numbers! When I authored this PR we were re running the extraction so I didn't have a sha yet to refer to. Given your feedback,now that we think we are done running the scripts I will pin it to a sha!

appilon committed:
Fix line numbers

Co-Authored-By: Nick Fagerlund <nick@hashicorp.com>

appilon commented on content/source/docs/extend/best-practices/deprecations.md:
```diff
@@ -73,7 +73,7 @@ The recommended process is as follows:
 1. Add a note to the documentation that either the existing (now the "old") attribute or new attribute must be configured.
 1. Add `ConflictsWith` to the schema definitions of both the old and new attributes so they will present an error to the operator if both are configured at the same time.
-1. Add conditional logic in the `Create`, `Read`, and `Update` functions of the data source or resource to handle both attributes. Generally, this involves using [`ResourceData.GetOk()`](https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceData.GetOk) (commonly `d.GetOk()` in HashiCorp maintained providers).
+1. Add conditional logic in the `Create`, `Read`, and `Update` functions of the data source or resource to handle both attributes. Generally, this involves using [`ResourceData.GetOk()`](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/schema#ResourceData.GetOk) (commonly `d.GetOk()` in HashiCorp maintained providers).
```
It won't be working until the SDK is a public GH repo

appilon committed:
pin links to commit

appilon committed:
create sdk migration page

appilon committed:
add link to providers section for sdk questions

kmoe committed:
add structure to standalone plugin SDK page

kmoe commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -0,0 +1,87 @@
+  Official standalone SDK for Terraform plugin development
+---
+
+As of September 2019, Terraform provider developers importing the Go module `github.com/hashicorp/terraform`, known as Terraform Core, should switch to `github.com/hashicorp/terraform-plugin-sdk`, the Terraform Plugin SDK, instead. 
```
Prior to public release we should update this if necessary. We could also add the relevant Core version.

kmoe commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -0,0 +1,87 @@
+
+The migration process can be automated with the [migrator tool](https://github.com/hashicorp/tf-sdk-migrator). This tool will check eligibility of a provider codebase, ensuring the use of Go modules and that the provider is upgrading from the latest `github.com/hashicorp/terraform` release. It will also identify if the handful of removed APIs are being used.
+
+You can also migrate your provider manually by replacing references to `github.com/hashicorp/terraform` with `github.com/hashicorp/terraform-plugin-sdk`. We recommend using the official migrator tool as it has a number of checks that will make this process safer. Please also read the deprecation notices below.
```
Add link to deprecations section.

kmoe commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -0,0 +1,87 @@
+
+If this command succeeds, proceed to Step 2.
+
+Otherwise, the tool will output the steps you need to take to ensure the provider can be migrated. Please see https://github.com/hashicorp/tf-sdk-migrator#check-eligibility-for-migration-tf-sdk-migrator-check for more information about the eligibility checks, and see the Deprecations section below for what to do if you are using deprecated packages or identifiers.
```
Add link to Deprecations section.

radeksimko commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -0,0 +1,87 @@
+
+# Why a separate module?
+
+While the `helper` package in Terraform Core has served us well, in order for provider development to evolve, the SDK needed to break out into its own repository. Terraform Core's versioning has been oriented towards practitioners. With the "unofficial" SDK existing in the core repository, the SDK becomes tied to Core releases and cannot follow semantic versioning. The new standalone SDK [github. ...
```
Nitpick - I'd write "`helper/*` and other packages" as it's not just one package and `helper` isn't even importable package.

appilon committed:
fill in deprecation details

kmoe commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -85,3 +85,11 @@ Congratulations! Your Terraform provider is migrated to the standalone SDK. You
 Version 1.0.0 of the standalone plugin SDK is intended to differ as little as possible from the legacy plugin SDK. However, we have had to deprecate some packages and identifiers.
 
 ## Deprecations
+The following packages, functions, and identifiers have been deprecated as of v0.12.7 of the legacy SDK in Core, and have removed altogether in the standalone SDK.
```
We should also describe deprecated packages (e.g. `flatmap`) here.

appilon committed:
finish documentation update
fix sidebar current for SDK page
note SDK in upgrading to v0.12 page
fix a few old SDK import references
mention flatmap removal

kmoe commented on content/source/docs/extend/plugin-sdk.html.md:
```diff
@@ -0,0 +1,99 @@
+  - `httpclient.UserAgentString()/terraform.UserAgentString()` have been removed. Please use [`httpclient.TerraformUserAgent()`](https://github.com/hashicorp/terraform-plugin-sdk/blob/e664f5b78081fde148c4ea55a0e068dc62fb2274/httpclient/useragent.go#L14) instead.
+  - `httpclient.New()` has been removed. Please use `github.com/hashicorp/go-cleanhttp.DefaultPooledClient()` directly with a custom transport. This is how [`httpclient.New()`](https://github.com/hashicorp/terraform/blob/39f61a07955b57c0a4afeb183259ca1697677148/httpclient/client.go#L12) was implemented in Core.
+
+  * A small number of providers have used the [flatmap package](https://github.com/hashicorp/terraform/tree/e1d0acda0b19be25ea96748896d3cd7117df955a/flatmap). Unfortunately it will not be apart of the standalone SDK. It is recommended to just move away from using that altogether, but as a quick solution, copy paste whatever is needed over and strip away the parts that are Terraform Core specific. ...
```
```suggestion
  * A small number of providers have used the [flatmap package](https://github.com/hashicorp/terraform/tree/e1d0acda0b19be25ea96748896d3cd7117df955a/flatmap). Unfortunately it will not be a part of the standalone SDK. It is recommended to just move away from using that altogether, but as a quick solution, copy paste whatever is needed over and strip away the parts that are Terraform Core specific. This package was never intended to be used outside of Terraform Core.
```

kmoe reviewed (approved):
LGTM with one typo

appilon committed:
fix typo

Co-Authored-By: kmoe <5575356+kmoe@users.noreply.github.com>
